promql/query_logger.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2013 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package promql
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"os"
+	"strings"
+	"syscall"
+	"time"
+)
+
+type ActiveQueryTracker struct {
+	mmapedFile   []byte
+	getNextIndex chan int
+	lastIndex    chan int
+	logger       log.Logger
+}
+
+const (
+	entrySize int = 1000
+)
+
+func parseBrokenJson(brokenJson []byte, logger log.Logger) (bool, string) {
+	queries := strings.TrimSpace(string(brokenJson))
+	queries = strings.Trim(queries, "\x00")
+	queries = strings.Trim(queries, " ")
+	queries = queries[:len(queries)-1] + "]"
+
+	// Conditional because of implementation detail: len() = 1 implies file consisted of a single char: '['.
+	if len(queries) == 1 {
+		return false, "[]"
+	}
+
+	var out bytes.Buffer
+	err := json.Indent(&out, []byte(queries), "", "\t")
+
+	if err != nil {
+		level.Warn(logger).Log("msg", "Error parsing json", "err", err)
+		return false, "[]"
+	}
+
+	return true, out.String()
+}
+
+func logUnfinishedQueries(filename string, filesize int, logger log.Logger) {
+	if _, err := os.Stat(filename); err == nil {
+		fd, err := os.Open(filename)
+		if err != nil {
+			level.Error(logger).Log("msg", "Failed to open query log file", "err", err)
+			return
+		}
+
+		brokenJson := make([]byte, filesize)
+		_, err = fd.Read(brokenJson)
+		if err != nil {
+			level.Error(logger).Log("msg", "Failed to read query log file", "err", err)
+			return
+		}
+
+		queriesExist, queries := parseBrokenJson(brokenJson, logger)
+		if !queriesExist {
+			return
+		}
+		level.Info(logger).Log("msg", "These queries didn't finish in prometheus' last run:")
+		fmt.Println(queries)
```

Comment:
log these normally

Reply:
Do you mean something like this?
```
level.Info(logger).Log("msg", "Some queries didn't finish in prometheus' last run", "queries", queries)
```

Because I am difficulty printing a formatted json with logger. Instead of true newlines and indents, I am getting a bunch of \n and \t mixed with rest of the json in my shell output.

Reply:
Done!